### PR TITLE
[CDLS-02] Add Authorization

### PR DESCRIPTION
**Pull request by klindziukp**
## PROBLEM
- Authorization and authentication is missed
## SOLUTION
- Enable `Spring Security` module
- Create `docker-compose` files
- Implement appropriate services
- Implement annotation for validations
## DEPENDS ON
- https://github.com/codeless-api-automation/codeless-common/pull/1
## TODO
- Add `Kanban` board for progress tracking
- Define `code-style` formatter
- Update permissions
- Discuss additional endpoints
- Update `login/logout` to custom flow

### CODE REVIEW

File: src/main/java/com/codeless/api/automation/security/controller/UserController.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    @Autowired
+    public UserController(UserServiceImpl userService) {
+        this.userService = userService;
+    }
+
+    @PostMapping(value = "/registration", produces = MediaType.APPLICATION_JSON_VALUE)
+    @JsonView(View.ALLOWED.class)
+    public ResponseEntity<User> register(@RequestBody @Valid User user) {
+        User createdUser = userService.saveUser(user);
+        return ResponseEntity.ok(createdUser);
```
Comment (Strunevskiy) [inline src/main/java/com/codeless/api/automation/security/controller/UserController.java:32]: Task, Oleg: Check usage of Valid and Hibernate.

File: src/main/java/com/codeless/api/automation/security/controller/UserController.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Autowired
+    public UserController(UserServiceImpl userService) {
+        this.userService = userService;
+    }
+
+    @PostMapping(value = "/registration", produces = MediaType.APPLICATION_JSON_VALUE)
+    @JsonView(View.ALLOWED.class)
+    public ResponseEntity<User> register(@RequestBody @Valid User user) {
+        User createdUser = userService.saveUser(user);
```
Comment (Strunevskiy) [inline src/main/java/com/codeless/api/automation/security/controller/UserController.java:31]: Update to `return userService.saveUser(user);`

File: src/main/java/com/codeless/api/automation/security/controller/UserController.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+
+@RestController
```
Comment (Strunevskiy) [inline src/main/java/com/codeless/api/automation/security/controller/UserController.java:18]: Add `@RequiredArgsConstructor` and remove constructor.

File: src/main/java/com/codeless/api/automation/security/domain/User.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import javax.persistence.Table;
+import java.util.Collection;
+import java.util.Collections;
+
+@Entity
+@Table(name = "users")
+@Data
+@Accessors(chain = true)
+@JsonView(View.ALLOWED.class)
+public class User implements UserDetails {
```
Comment (Strunevskiy) [inline src/main/java/com/codeless/api/automation/security/domain/User.java:27]: split user into DTO and Entity.

File: src/main/java/com/codeless/api/automation/security/repository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.codeless.api.automation.security.repository;
+
+import com.codeless.api.automation.security.domain.User;
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (Strunevskiy) [inline src/main/java/com/codeless/api/automation/security/repository/UserRepository.java:7]: Move to common module.

File: src/main/java/com/codeless/api/automation/security/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.Objects;
+
+@Service
+public class UserServiceImpl implements UserService, UserDetailsService {
+
+    private static final Logger logger = LogManager.getLogger(UserServiceImpl.class);
```
Comment (Strunevskiy) [inline src/main/java/com/codeless/api/automation/security/service/impl/UserServiceImpl.java:21]: Please replace Logger with @Slf4j.

File: src/main/java/com/codeless/api/automation/security/controller/UserController.java (hunk 1 above)
Comment (Strunevskiy) [reply src/main/java/com/codeless/api/automation/security/controller/UserController.java:32]: Done!

**Review by Strunevskiy — APPROVED** [review]
Once comments are addressed, let's merge this PR. Also, as of now, all the endpoint could be made public until UI has registration page as well as login page.

File: src/main/java/com/codeless/api/automation/security/controller/UserController.java (hunk 2 above)
Comment (klindziukp) [reply src/main/java/com/codeless/api/automation/security/controller/UserController.java]: done by Oleg.

File: src/main/java/com/codeless/api/automation/security/domain/User.java (hunk 4 above)
Comment (klindziukp) [reply src/main/java/com/codeless/api/automation/security/domain/User.java]: splitter and moved to common

File: src/main/java/com/codeless/api/automation/security/repository/UserRepository.java (hunk 5 above)
Comment (klindziukp) [reply src/main/java/com/codeless/api/automation/security/repository/UserRepository.java]: moved

**Review by Strunevskiy — APPROVED** [review]

**Pull request merged**